### DISCUSSION TITLE
How to read and write json file

### DISCUSSION THREAD

**Discussion by ghost** [nlohmann/json#2808] [Q&A] [answered]
Hello! I am new and learning C++ for the first time and want to store data from my C++ application to a .json file and read the data again in C++.

My example .json file:
```json
{
    "Name1":
    {
        "Food":"Candy",
        "Number":100
    },
    "Name2":
    {
        "Food":"Cake",
        "Number":null
    }
}
```
How could I read the json file and store Name1's Food (Candy) as a string in C++? Along with the Name1's Number and Name2's items.

And how could I write/edit the json file and save it? So that I can change Name1's Food to something like cookies?

And last but not least, how could I delete a part of the json file in C++? For example, if I want to delete the Food id or the Name2.

**Comment by nlohmann** [comment, marked as answer]
Assuming your file is called `data.json`, this should work:

```cpp
// parse
std::ifstream i("data.json");
json j;
i >> j;

// read
std::string food = j["Name1"]["Food"];

// change
j["Name1"]["Food"] = "Cookie";

// delete
j.erase("Name1");
```

See https://github.com/nlohmann/json and https://json.nlohmann.me for more information.

**Reply by ghost** [reply]
Thank you so much! This is really helpful and I love your repo!

**Reply by ghost** [reply]
I have one more problem @nlohmann , how can I save the j back to the data.json so that my changes are saved?
I have tried something your website said:
```json
i << std::setw(4) << j << std::endl;
```
but the compiler tells me to use getw(4) not setw(4) and when I do that I get a bunch of errors.

**Reply by nlohmann** [reply]
You defined i as input file (std::ifstream). You need a std::ofstream to write.

**Comment by ghost** [comment]
Ah thank you @nlohmann , that solved my problem! Another problem I got working on my project was that when I try to delete something inside of something like for example deleting number inside of Name2 I found no way to do so.
I tried j.erase(["Name2"],["Number"]) and j.erase("Name2","Number") and even j.erase["Name2"]["Number"]

Another, (and hopefully last thing) is if your json.hpp has a function to delete the contents of a file, I ask this because when I save it, it writes next to the data that was previously there, I want to overwrite the data. If not is there a way to delete the contents of a file in C++? I have defined i as input and output using #include <fstream> for reading AND writing.

Your help is much appreciated!

**Reply by ghost** [reply]
I was able to fix my problem with saving the file by reading the file with ifstream and setting that to j then writing with ofstream. It was able to overwrite the data in the file with the new data.
I still have my problem with deleting tho.

**Reply by nlohmann** [reply]
See https://json.nlohmann.me/api/basic_json/erase/ for documentation and examples how to erase values.
